### PR TITLE
Add hook to change metadata

### PR DESCRIPTION
**Pull request by andreasschacht**

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -26,14 +26,26 @@ function defaultConfig() {
...
+   * This hook will be called after the metaData generation in the postprocessTree-hook.
+   * If you need to change or extend the metaData, you can implement this method in your addon.
+   *
+   * @method addMetaData
+   * @param {Object} metaData
+   * @return {Object} the changed metaData
+   * @public
+   */
+  addMetaData(metaData) {
+    return metaData;
```
Comment (simonihmig) [inline index.js]: Your intention is that another addon overrides (patches) this method? Seems a bit weird, that another object is patching this one. Also does not allow for multiple addons to hook into the process...

Maybe something like `addMetaDataHook(fn)`, that adds functions to an internal array, and processes them iteratively:
```js
processMetaData(metaData) {
  this._hooks.forEach(fn => metaData = fn(metaData));
  return metaData;
}
```

Comment (andreasschacht) [reply index.js]: Sure, you're right. That was also my first approach... see 2 commit before...
For some reason i thought, it looks cleaner this way...

Comment (simonihmig) [reply index.js]: Ah, yes, nice. I like that commit better ;)

`target` support is nice, but not really required I think, as nowadays you can easily "fix" the `this` scope, using arrow functions or `Function#bind()`. The implementation can be simplified without it. But it's fine either way!

**Comment by simonihmig** [comment]
👆 Well, except for the failing tests 😬

**Comment by andreasschacht** [comment]
@dajk @simonihmig I added some major changes, now you can hook into the image process. So please re-Review ;-)

File: README.md [hunk 2]
```diff
@@ -211,6 +211,49 @@ This mixin binds the url of the best fitting image to the source attribute, base
 
 This mixin binds the url of the best fitting image as the background url to the elements style attribute, based in the values provided by the `image` and `size` attribute. It also get the `responsiveImage` service injected.
 
+##Extensibility hooks
```
Comment (simonihmig) [inline README.md]: this does not render correctly, because of the missing space!

File: README.md [hunk 3]
```diff
@@ -211,6 +211,49 @@ This mixin binds the url of the best fitting image to the source attribute, base
...
+### Extend the image processing
+
+During the image process the addon calls the `preProcessImage` and the `postProcessImage` hooks for each origin image and supported width. Here you can add custom image process steps, like a watermark integration. The first hook will be called just before the addon's image process calls applies, the latter after. You can register your callbacks by calling the addon's `addImagePreProcessor` or `addImagePostProcessor` function before the addon's `postprocessTree` was called.
+In both cases the callback method you provide must have the following signature:
```
Comment (simonihmig) [inline README.md]: I think an example how to get access to the addon instance would be useful!

File: README.md [hunk 4]
```diff
@@ -211,6 +211,49 @@ This mixin binds the url of the best fitting image to the source attribute, base
...
+  function customMetadata(image, metadata, configuration)
+  { 
+    return metadata;
+  }
+```
+* **image:** the name of the origin image file
+* **metadata:** object with the metadata of the generated images
+* **configuration:** the configuration for the image generation (from environments configuration)
+
+The callback must return an object with the extended metadatas.
```
Comment (simonihmig) [inline README.md]: just "metadata", data is already plural ;)

File: README.md [hunk 5]
```diff
@@ -211,6 +211,49 @@ This mixin binds the url of the best fitting image to the source attribute, base
...
+
+The callback must return a `sharp`-object or a Promise resolves to it.
+
+**Note:** In addition to the callback, you can also pass an optional target object that will be set as `this` on the context. This is a good way to give your function access to the current object.
+
+For an example see [ember-lazy-responsive-image](https://github.com/kaliber5/ember-lazy-responsive-image/blob/master/index.js)
+### Extend the metadata
+
+Before the addon injects the generated metadata into the build, a `extendMetadata`-hook is called for each origin image. Here you can add custom metadata.  You can register your callbacks by calling the addon's `addMetadataExtension` function before the addon's `postprocessTree` was called.
+The callback method you provide must have the following signature:
```
Comment (simonihmig) [inline README.md]: method -> function

File: README.md [hunk 6]
```diff
@@ -211,6 +211,49 @@ This mixin binds the url of the best fitting image to the source attribute, base
...
+* **configuration:** the configuration for the current image processing (from environments configuration)
+
+The callback must return a `sharp`-object or a Promise resolves to it.
+
+**Note:** In addition to the callback, you can also pass an optional target object that will be set as `this` on the context. This is a good way to give your function access to the current object.
+
+For an example see [ember-lazy-responsive-image](https://github.com/kaliber5/ember-lazy-responsive-image/blob/master/index.js)
+### Extend the metadata
+
+Before the addon injects the generated metadata into the build, a `extendMetadata`-hook is called for each origin image. Here you can add custom metadata.  You can register your callbacks by calling the addon's `addMetadataExtension` function before the addon's `postprocessTree` was called.
```
Comment (simonihmig) [inline README.md]: Maybe add a sentence for context, that this metadata is what is returned by the service's `getImages ` and `getImageDataBySize` methods?

File: README.md (hunk 3 above)
Comment (andreasschacht) [reply README.md]: A few lines later I point to the new `ember-lazy-responsive-image` - addon as an example. For now it's empty. But the addon should be ready 'till the next days.

File: README.md (hunk 6 above)
Comment (andreasschacht) [reply README.md]: I added a sentence and also point to the `ember-lazy-responsive-image`-addon as an example. Do you think it's enough?

File: README.md (hunk 2 above)
Comment (andreasschacht) [reply README.md]: Fixed. For some reason the editor's plugin renders correctly...

File: README.md (hunk 5 above)
Comment (andreasschacht) [reply README.md]: Fixed

File: README.md (hunk 4 above)
Comment (andreasschacht) [reply README.md]: Fixed

File: README.md (hunk 6 above)
Comment (simonihmig) [reply README.md]: Sure

**Pull request merged**